modules/ding_wayf_dk/modules/ding_wayf_dk_login/ding_wayf_dk_login.module

```diff
@@ -0,0 +1,255 @@
+    '#title' => t('Login with NemID'),
+  );
+
+  $variables['login_url']['#text'] = t('Login with') . ' ' . drupal_render($image);
+}
+
+/**
+ * Helper function to redirect the user.
+ *
+ * If destination query string exists the user is redirect to it else the
+ * fallback URL given is used.
+ *
+ * @param string $fallback_url
+ *   URL used if the destination query sting is not set.
+ */
+function ding_wayf_dk_login_redirect_user($fallback_url = NULL) {
+  if (!empty($_REQUEST['destination'])) {
+    // We do not user drupal_get_destination here as it will set the destination
+    // if one do not exists.
+    $fallback_url = $_REQUEST['destination'];
+  }
+  else {
+    global $user;
+    $fallback_url = token_replace($fallback_url, array('user' => $user));
+  }
+
+  drupal_goto($fallback_url);
+}
+
+/**
+ * Implements hook_form_wayf_dk_login__settings_form_alter().
+ *
+ * Adds debug option to map test WAYF CPR number to a provider valid one.
+ */
+function ding_wayf_dk_login_form_wayf_dk_login__settings_form_alter(&$form, &$form_state, $form_id) {
+  $form['settings']['debug'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Ding WAYF debug'),
+  );
+
+  $form['settings']['debug']['ding_wayf_dk_login_debug'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Enable debug'),
+    '#default_value' => variable_get('ding_wayf_dk_login_debug', FALSE),
+    '#description' => t('Enable debug to do CPR mapping with WAYF test users.'),
+  );
+
+  $form['settings']['debug']['ding_wayf_dk_login_debug_cpr'] = array(
+    '#type' => 'password',
+    '#title' => t('CPR map'),
+    '#description' => t('Map student test CPR (0708741234).'),
+    '#default_value' => variable_get('ding_wayf_dk_login_debug_cpr', ''),
+  );
+}
+
+/**
+ * Get user credentials from SAML attributes and the user provider.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ *
+ * @return array
+ *   Attributes filter based on the active provider. Defaults to
+ *   'schacPersonalUniqueID' with social security number (CPR).
+ */
+function ding_wayf_dk_login_get_credentials(array $attributes) {
+  // Set default required attributes.
+  $required_attributes = array(
+    'schacPersonalUniqueID' => array(
+      'field' => 'CPR',
+      'authname' => TRUE,
+    ),
+  );
+
+  // Ask the provider about required attributes and give it a change to override
+  // defaults.
+  if (ding_provider_implements('wayf', 'attributes')) {
+    $required_attributes = ding_provider_invoke('wayf', 'attributes', $required_attributes);
+  }
+
+  // Filter the attributes from WAYF/SAML and only return required once.
+  return ding_wayf_dk_login_get_required_attributes($attributes, $required_attributes);
+}
+
+/**
+ * Extract required attributes from available WAYF attributes.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ * @param array $required_attributes
+ *   Required attributes.
+ *
+ * @return array
+ *   Will be empty if not all required attributes is present.
```

Comment:
This does not seem to hold true? If there are two required attributes and the attributes only contain one of them I cannot spot anywhere in the function which will lead to it returning an empty array.